### PR TITLE
feat(resharding): flat storage resharding children catchup

### PR DESCRIPTION
**Pull request by Trisfald**
PR to add children catchup step for flat storages created as a result of a parent shard split.

In previous iterations, the two children shards were populated in a background task from the flat storage of the parent at height `last block of old shard layout` (post-processing).

Since the task mentioned above takes a long time and the children are active shards in the `first block of the new shard layout` their flat storage accumulates a lot of deltas.

The catchup step applies delta in the background, then finalizes creation of child flat storage, and triggers a possible memtrie rebuild.

Part of #12174

### CODE REVIEW

File: chain/chain/src/resharding/resharding_actor.rs [hunk 1]
```diff
@@ -1,23 +1,62 @@
...
+        &self,
+        resharder: FlatStorageResharder,
+        shard_uid: ShardUId,
+        flat_head_block_hash: CryptoHash,
+    ) {
+        if matches!(
+            resharder.shard_catchup_task(shard_uid, flat_head_block_hash, &self.chain_store),
+            FlatStorageReshardingTaskStatus::Failed
+        ) {
+            panic!("impossible to recover from a flat storage resharding failure");
```
Comment (Trisfald) [inline chain/chain/src/resharding/resharding_actor.rs:59]: Quick and dirty. It can be changed later to something meaningful

File: chain/chain/src/resharding/resharding_actor.rs [hunk 2]
```diff
@@ -1,23 +1,62 @@
...
 
 impl ReshardingActor {
-    pub fn new() -> Self {
-        Self {}
+    pub fn new(store: Store, genesis_height: BlockHeight) -> Self {
+        Self { chain_store: ChainStore::new(store, genesis_height, false) }
+    }
+
+    fn handle_memtrie_reload(&self, _shard_uid: ShardUId) {
+        // TODO(resharding)
```
Comment (Trisfald) [inline chain/chain/src/resharding/resharding_actor.rs:40]: I thought this may be the entrypoint for memtrie rebuild.

File: chain/chain/src/flat_storage_resharder.rs [hunk 3]
```diff
@@ -1514,4 +1765,200 @@ mod tests {
         );
         assert_eq!(flat_store.get(right_child_shard, &buffered_receipt_key), Ok(None));
     }
+
+    /// Base test scenario for testing children catchup.
+    fn children_catchup_base(with_restart: bool) {
```
Comment (Trisfald) [inline chain/chain/src/flat_storage_resharder.rs:1770]: The most relevant change in tests is .. the addition of this test.

File: chain/chain/src/flat_storage_resharder.rs [hunk 4]
```diff
@@ -460,6 +476,146 @@ impl FlatStorageResharder {
...
+                }
+                flat_head_block_hash = chain_store.get_next_block_hash(&flat_head_block_hash)?;
+                if let Some(changes) = store
+                    .get_delta(shard_uid, flat_head_block_hash)
+                    .map_err(|err| Into::<StorageError>::into(err))?
+                {
+                    merged_changes.merge(changes);
+                    store_update.remove_delta(shard_uid, flat_head_block_hash);
+                }
+                // TODO (resharding): if flat_head_block_hash == state sync hash -> do snapshot
```
Comment (Trisfald) [inline chain/chain/src/flat_storage_resharder.rs:554]: @marcelo-gonzalez This one is the designated point to trigger the snapshot. We can discuss how it can be done in practice, the important consideration is that here the flat storage has all the state needed.

File: chain/chain/src/flat_storage_resharder.rs [hunk 5]
```diff
@@ -357,8 +364,13 @@ impl FlatStorageResharder {
         let (parent_shard, split_status) = self
             .get_parent_shard_and_status()
             .expect("flat storage resharding event must be Split!");
-        let SplittingParentStatus { left_child_shard, right_child_shard, flat_head, .. } =
-            split_status;
+        let SplittingParentStatus {
+            left_child_shard,
+            right_child_shard,
+            flat_head,
+            block_hash,
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:371]: nit: What is this block hash? Can you rename it to something more telling? Maybe something similar to sync hash in state sync like `resharding_hash` that is well defined?

File: chain/chain/src/flat_storage_resharder.rs [hunk 6]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+        // If flat storage current status is beyond the chain final block, try again later.
+        let chain_final_head = chain_store.final_head().unwrap();
+        let height_result = chain_store.get_block_height(&flat_head_block_hash);
+        let Ok(height) = height_result else {
+            error!(target: "resharding", ?shard_uid, ?flat_head_block_hash, err = ?height_result.unwrap_err(), "can't find flat head height!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        if height > chain_final_head.height {
+            info!(target = "resharding", ?height, chain_final_height = ?chain_final_head.height, "flat head beyond chain final tip: postponing flat storage shard catchup task");
+            self.scheduler.send(ReshardingRequest::FlatStorageShardCatchup {
```
Comment (shreyan-gupta) [inline chain/chain/src/flat_storage_resharder.rs:498]: Wait, if this check is saying height is already beyond final head, why are we calling the schedule here?

File: chain/chain/src/flat_storage_resharder.rs [hunk 7]
```diff
@@ -381,11 +393,15 @@ impl FlatStorageResharder {
                     store_update.set_flat_storage_status(
                         child_shard,
                         FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CatchingUp(
-                            flat_head.hash,
+                            block_hash,
                         )),
                     );
+                    self.scheduler.send(ReshardingRequest::FlatStorageShardCatchup {
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:399]: Is this going from the resharding actor to itself?

File: chain/chain/src/flat_storage_resharder.rs [hunk 8]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+    ) -> FlatStorageReshardingTaskStatus {
+        info!(target: "resharding", ?shard_uid, ?flat_head_block_hash, "flat storage shard catchup task started");
+        // If flat storage current status is beyond the chain final block, try again later.
+        let chain_final_head = chain_store.final_head().unwrap();
+        let height_result = chain_store.get_block_height(&flat_head_block_hash);
+        let Ok(height) = height_result else {
+            error!(target: "resharding", ?shard_uid, ?flat_head_block_hash, err = ?height_result.unwrap_err(), "can't find flat head height!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        if height > chain_final_head.height {
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:496]: When can this happen?

File: chain/chain/src/flat_storage_resharder.rs [hunk 9]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+            self.scheduler.send(ReshardingRequest::FlatStorageShardCatchup {
+                resharder: self.clone(),
+                shard_uid,
+                flat_head_block_hash,
+            });
+            return FlatStorageReshardingTaskStatus::Cancelled;
+        }
+
+        // If the flat head is not in the canonical chain this task has failed.
+        match chain_store.get_block_hash_by_height(height) {
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:507]: That isn't what I expected the check for canonical to look like but alright :)

File: chain/chain/src/flat_storage_resharder.rs [hunk 10]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+        let store = self.runtime.store().flat_store();
+        let mut store_update = store.store_update();
+        // Deltas must exist because we applied them previously.
+        let deltas_metadata = store.get_all_deltas_metadata(shard_uid).unwrap_or_else(|_| {
+            panic!("Cannot read flat state deltas metadata for shard {shard_uid} from storage")
+        });
+        let mut deltas_gc_count = 0;
+        for delta_metadata in deltas_metadata {
+            if delta_metadata.block.height <= flat_head.height {
+                store_update.remove_delta(shard_uid, delta_metadata.block.hash);
```
Comment (shreyan-gupta) [inline chain/chain/src/flat_storage_resharder.rs:631]: I would like @Longarithm to take a look at this function and check whether it looks good

File: chain/chain/src/resharding/resharding_actor.rs [hunk 11]
```diff
@@ -1,23 +1,77 @@
...
-impl Handler<FlatStorageSplitShardRequest> for ReshardingActor {
-    fn handle(&mut self, msg: FlatStorageSplitShardRequest) {
-        self.handle_flat_storage_split_shard_request(msg);
+impl HandlerWithContext<ReshardingRequest> for ReshardingActor {
+    fn handle(&mut self, msg: ReshardingRequest, ctx: &mut dyn DelayedActionRunner<Self>) {
+        match msg {
+            ReshardingRequest::FlatStorageSplitShard { resharder } => {
+                self.handle_flat_storage_split_shard(resharder);
+            }
+            ReshardingRequest::FlatStorageShardCatchup {
```
Comment (shreyan-gupta) [inline chain/chain/src/resharding/resharding_actor.rs:24]: These are three different types of requests. We should not group them under `ReshardingRequest` but instead separate these into three different Messages.

**Review by shreyan-gupta — COMMENTED** [review]
Would love others to take a look as well

File: chain/chain/src/flat_storage_resharder.rs [hunk 12]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+    /// Applies flat storage deltas in batches on a shard that is in catchup status.
+    ///
+    /// Returns the number of delta batches applied and the final tip of the flat storage.
+    fn shard_catchup_apply_deltas(
+        &self,
+        shard_uid: ShardUId,
+        mut flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> Result<(usize, Tip), Error> {
+        const CATCH_UP_BLOCKS: u32 = 50;
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:551]: Can you make it be part of the config? 

Cool that we have rate limiting for this. Why number of blocks? It makes perfect sense but I'm just curious what's your thinking.

File: chain/chain/src/flat_storage_resharder.rs [hunk 13]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+
+        let mut num_batches_done: usize = 0;
+
+        loop {
+            let _span = tracing::debug_span!(
+                target: "resharding",
+                "shard_catchup_apply_deltas/batch",
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:558]: Is `/` a supported character in the tracing span names?

File: chain/chain/src/flat_storage_resharder.rs [hunk 14]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+            let store = self.runtime.store().flat_store();
+            let mut store_update = store.store_update();
+
+            // Merge deltas from the next blocks until we reach chain final head.
+            for _ in 0..CATCH_UP_BLOCKS {
+                let height = chain_store.get_block_height(&flat_head_block_hash)?;
+                if height > chain_final_head.height {
+                    let msg = format!("flat head moved too far: head = {flat_head_block_hash}, height = {height}, final block height = {}", chain_final_head.height);
+                    return Err(Error::ReshardingError(msg));
+                }
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:574]: Can this happen or is it just in case?

File: chain/chain/src/flat_storage_resharder.rs [hunk 15]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+                    break;
+                }
+                flat_head_block_hash = chain_store.get_next_block_hash(&flat_head_block_hash)?;
+                if let Some(changes) = store
+                    .get_delta(shard_uid, flat_head_block_hash)
+                    .map_err(|err| Into::<StorageError>::into(err))?
+                {
+                    merged_changes.merge(changes);
+                    store_update.remove_delta(shard_uid, flat_head_block_hash);
+                }
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:586]: I would feel a bit better if this was using some flat storage API rather than reading directly from db.

File: chain/chain/src/flat_storage_resharder.rs [hunk 16]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+
+        loop {
+            let _span = tracing::debug_span!(
+                target: "resharding",
+                "shard_catchup_apply_deltas/batch",
+                ?shard_uid,
+                ?flat_head_block_hash,
+                batch_id = ?num_batches_done)
+            .entered();
+            let chain_final_head = chain_store.final_head()?;
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:563]: Cool that you refresh this from within the loop, it will get as close to final as possible.

File: chain/chain/src/flat_storage_resharder.rs [hunk 17]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+                // TODO (resharding): if flat_head_block_hash == state sync hash -> do snapshot
+            }
+
+            // Commit all changes to store.
+            merged_changes.apply_to_flat_state(&mut store_update, shard_uid);
+            store_update.set_flat_storage_status(
+                shard_uid,
+                FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CatchingUp(
+                    flat_head_block_hash,
+                )),
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:596]: Is this designed for resuming from the last delta if catch up was interrupted?

File: chain/chain/src/flat_storage_resharder.rs [hunk 18]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+        let deltas_metadata = store.get_all_deltas_metadata(shard_uid).unwrap_or_else(|_| {
+            panic!("Cannot read flat state deltas metadata for shard {shard_uid} from storage")
+        });
+        let mut deltas_gc_count = 0;
+        for delta_metadata in deltas_metadata {
+            if delta_metadata.block.height <= flat_head.height {
+                store_update.remove_delta(shard_uid, delta_metadata.block.hash);
+                deltas_gc_count += 1;
+            }
+        }
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:634]: Can you do it as part of applying the deltas? It would be more clear that those are already applied.

File: chain/chain/src/flat_storage_resharder.rs [hunk 19]
```diff
@@ -1514,4 +1799,200 @@ mod tests {
...
+                .save_flat_state_changes(
+                    block_hash,
+                    prev_hash,
+                    height,
+                    left_child_shard,
+                    &state_changes_left_child,
+                )
+                .unwrap()
+                .commit()
+                .unwrap();
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:1864]: nit: This is copy pasted, can you refactor to a method?

File: chain/chain/src/resharding/resharding_actor.rs [hunk 20]
```diff
@@ -1,23 +1,77 @@
...
+                resharder,
+                shard_uid,
+                flat_head_block_hash,
+            } => {
+                // Shard catchup task is delayed and could get postponed several times. This must be
+                // done to cover the scenario in which catchup is triggered so fast that the initial
+                // state of the new flat storage is beyond the chain final tip.
```
Comment (wacban) [inline chain/chain/src/resharding/resharding_actor.rs:31]: omg, thanks for the comment

File: chain/chain/src/resharding/resharding_actor.rs [hunk 21]
```diff
@@ -1,23 +1,77 @@
...
+    }
+
+    fn handle_memtrie_reload(&self, _shard_uid: ShardUId) {
+        // TODO(resharding)
+    }
+
+    fn handle_flat_storage_split_shard(&self, resharder: FlatStorageResharder) {
+        if matches!(resharder.split_shard_task(), FlatStorageReshardingTaskStatus::Failed) {
+            panic!("impossible to recover from a flat storage resharding failure");
+        }
```
Comment (wacban) [inline chain/chain/src/resharding/resharding_actor.rs:61]: nit: Can you make it into a match with all variants handled explicitly? If nothing is to be done just leave a comment.

File: chain/chain/src/resharding/resharding_actor.rs (hunk 1 above)
Comment (wacban) [reply chain/chain/src/resharding/resharding_actor.rs:74]: Can you differentiate the two panic with different messages?

File: chain/chain/src/resharding/types.rs [hunk 22]
```diff
@@ -1,19 +1,29 @@
 use crate::flat_storage_resharder::FlatStorageResharder;
 use near_async::messaging::Sender;
+use near_primitives::hash::CryptoHash;
+use near_store::ShardUId;
 
-/// Represents a request to start the split of a parent shard flat storage into two children flat
-/// storages.
+/// Represents a request to start a flat storage resharding task.
```
Comment (wacban) [inline chain/chain/src/resharding/types.rs:6]: Now it represents requests to start various stages of resharding

**Review by wacban — COMMENTED** [review]
Looks good at a first glance, I added a few comments for now.

File: chain/chain/src/flat_storage_resharder.rs (hunk 7 above)
Comment (shreyan-gupta) [reply chain/chain/src/flat_storage_resharder.rs:399]: It becomes really really hard to figure out which of these functions are called from within resharding actor and which aren't. That's why I wanted us to keep all functions executed in the actor as part of the actor instead of here.

Comment (shreyan-gupta) [reply chain/chain/src/flat_storage_resharder.rs:399]: Even just for this review, I had to bounce around several times to see who is calling who and when does a resharding request get converted to a function call in flat_storage_resharder

Comment (wacban) [reply chain/chain/src/flat_storage_resharder.rs:399]: Yeah now that you mentioned it I also struggled to grasp what's happening where. Do you think this should be addressed in this PR or as a follow up? @shreyan-gupta @Trisfald

File: chain/chain/src/flat_storage_resharder.rs (hunk 9 above)
Comment (wacban) [reply chain/chain/src/flat_storage_resharder.rs:507]: Oh I don't know, it just surprised me. There is a function call `is_on_current_chain` in chain that seems to work in a similar way.

File: chain/chain/src/flat_storage_resharder.rs [hunk 23]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
         Ok(iter)
     }
+
+    /// Task to perform catchup and creation of a flat storage shard spawned from a previous
+    /// resharding operation. This may be a long operation time-wise.
+    pub fn shard_catchup_task(
+        &self,
+        shard_uid: ShardUId,
+        flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:486]: I'd love to not have to pass this chain_store thingy around 🤣

File: chain/chain/src/flat_storage_resharder.rs [hunk 24]
```diff
@@ -58,7 +60,7 @@ pub struct FlatStorageResharder {
     /// The current active resharding event.
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:60]: Done!

File: chain/chain/src/resharding/resharding_actor.rs [hunk 25]
```diff
@@ -1,23 +1,77 @@
...
+    pub fn new(store: Store, genesis_height: BlockHeight) -> Self {
+        Self { chain_store: ChainStore::new(store, genesis_height, false) }
+    }
+
+    fn handle_memtrie_reload(&self, _shard_uid: ShardUId) {
+        // TODO(resharding)
+    }
+
+    fn handle_flat_storage_split_shard(&self, resharder: FlatStorageResharder) {
+        if matches!(resharder.split_shard_task(), FlatStorageReshardingTaskStatus::Failed) {
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs:59]: Done!

File: chain/chain/src/resharding/resharding_actor.rs (hunk 1 above)
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs:74]: Done!

File: chain/chain/src/flat_storage_resharder.rs [hunk 26]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+                    store_update.remove_delta(shard_uid, flat_head_block_hash);
+                }
+                // TODO (resharding): if flat_head_block_hash == state sync hash -> do snapshot
+            }
+
+            // Commit all changes to store.
+            merged_changes.apply_to_flat_state(&mut store_update, shard_uid);
+            store_update.set_flat_storage_status(
+                shard_uid,
+                FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CatchingUp(
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:598]: We set status to CatchingUp the first time in `split_shard_task_postprocessing` after all key values have been split successfully.

Then we do it in each batch to update the block has of the flat head.

File: chain/chain/src/resharding/resharding_actor.rs (hunk 21 above)
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs]: Done!

File: chain/chain/src/flat_storage_resharder.rs (hunk 12 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:555]: It's the same logic used in `flat_storage_creator` to limit memory usage.

Sure, I can add it in the config, likely in another PR.

File: chain/chain/src/flat_storage_resharder.rs (hunk 5 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:375]: Done!

File: chain/chain/src/flat_storage_resharder.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:553]: I'll upgrade CATCH_UP_BLOCKS to a config param as suggested in another comment

File: chain/chain/src/flat_storage_resharder.rs [hunk 27]
```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
...
+
+            // Merge deltas from the next blocks until we reach chain final head.
+            for _ in 0..CATCH_UP_BLOCKS {
+                let height = chain_store.get_block_height(&flat_head_block_hash)?;
+                if height > chain_final_head.height {
+                    let msg = format!("flat head moved too far: head = {flat_head_block_hash}, height = {height}, final block height = {}", chain_final_head.height);
+                    return Err(Error::ReshardingError(msg));
+                }
+                // Stop if we reached chain final head.
+                if flat_head_block_hash == chain_final_head.last_block_hash {
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:578]: Is it impossible even in presence of forks? I don't have a test yet so this part is still fuzzy in my head.
(Admittedly I copied this check from catchup of flat storage creation)

File: chain/chain/src/flat_storage_resharder.rs (hunk 6 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:500]: Oh this is a surprise I got from test loop resharding test.
It shouldn't happen in practice but the logic is the following: 
If the split operation is very fast, since we split parent flat head and deltas together, the children flat storage after the split may have their flat head beyond the final block height. We don't want that because of the invariant that flat head <= final block height, so we delay the catchup until the canonical chain contains the children flat head hash.

File: chain/chain/src/resharding/resharding_actor.rs (hunk 11 above)
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs:24]: I took inspiration from other requests that are using `enum` to group requests. 
I can make three different requests and group the resulting three senders into a struct. Can you explain what would be the advantage compared to this approach?

File: chain/chain/src/flat_storage_resharder.rs (hunk 8 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:498]: Answered in Shreyan's comment!

File: chain/chain/src/flat_storage_resharder.rs (hunk 13 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:560]: Logs work fine. I couldn't find any reference to forbidden characters.. so should be ok

File: chain/chain/src/flat_storage_resharder.rs (hunk 15 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:588]: I don't think we can use flat storage API (if it exits) since the flat storage for this child doesn't yet exist in flat storage manager

File: chain/chain/src/flat_storage_resharder.rs (hunk 17 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:598]: Correct

File: chain/chain/src/flat_storage_resharder.rs (hunk 19 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:1864]: Done!

File: chain/chain/src/resharding/resharding_actor.rs [hunk 28]
```diff
@@ -1,23 +1,77 @@
...
+            }
+            ReshardingRequest::FlatStorageShardCatchup {
+                resharder,
+                shard_uid,
+                flat_head_block_hash,
+            } => {
+                // Shard catchup task is delayed and could get postponed several times. This must be
+                // done to cover the scenario in which catchup is triggered so fast that the initial
+                // state of the new flat storage is beyond the chain final tip.
+                ctx.run_later(
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs:32]: Maybe `run_later` is misused here. This has been my tentative solution to the following problem.

Originally, I was calling directly `handle_flat_storage_shard_catchup` and waiting some time inside the function. But it was failing in test loop for resharding_v3 because the test was getting stuck inside this function call. The other actors weren't progressing at all and the condition to resume catchup wasn't satisfied ever.

Basically I need a way to postpone execution of catchup until the canonical chain makes enough progress. I found that `run_later` achieves that by allowing other actors to make progress in the meantime. 

Should I rather change the test_loop test to avoid this problem?
Please feel free to suggest a better approach (even offline)

File: chain/chain/src/flat_storage_resharder.rs (hunk 14 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:576]: I'm afraid this can happen in the presence of forks. To be confirmed with a resharding-with-forks test..

File: chain/chain/src/flat_storage_resharder.rs (hunk 18 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:636]: It somehow feels safer to remove deltas only when we know for sure the flat storage is going to be created.

File: chain/chain/src/resharding/resharding_actor.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (shreyan-gupta) [reply chain/chain/src/resharding/resharding_actor.rs:32]: [This](https://github.com/near/nearcore/blob/e6dd1a447a950ae3716b9d682e3c305cf715d86b/chain/client/src/sync_jobs_actor.rs) is an example of how run_later is used. We can discuss this offline

File: chain/chain/src/flat_storage_resharder.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (shreyan-gupta) [reply chain/chain/src/flat_storage_resharder.rs:578]: flat head height in the beginning should always be <= final_head height. If we maintain that constraint, then the flat head height of the children should also be <= final_head height.

Final head is sorta defined as "definitely part of blockchain, without forks"

File: chain/chain/src/resharding/resharding_actor.rs (hunk 11 above)
Comment (shreyan-gupta) [reply chain/chain/src/resharding/resharding_actor.rs:24]: Yeah, no, it's just additional boilerplate code at the call site and at the end of the day all requests are effectively different. We've already defined them to be part of resharding as we are messaging to the resharding actor.

We've made this mistake a long time ago in ShardsManager already (NetworkRequest is still fine but I would ideally like to separate NetworkRequest from PeerManagerAdapter too). Anyway, I don't think we follow this enum pattern for new code.

File: chain/chain/src/flat_storage_resharder.rs (hunk 6 above)
Comment (shreyan-gupta) [reply chain/chain/src/flat_storage_resharder.rs:500]: Hmm, I see, well, I don't think this should be the solution. Instead what we should do is delay and retry the original split flat storage request (the run_later logic) for FlatStorageSplitShard and then we can hopefully remove this code and the retry logic for `FlatStorageShardCatchup` (assuming FlatStorageShardCatchup is called AFTER split is completed)

File: chain/chain/src/flat_storage_resharder.rs (hunk 14 above)
Comment (shreyan-gupta) [reply chain/chain/src/flat_storage_resharder.rs:576]: Stated in comment [here](https://github.com/near/nearcore/pull/12312#discussion_r1820701423), I guess we can avoid forks processing etc. altogether if we start our flat storage split AFTER split height <= chain head

File: chain/chain/src/flat_storage_resharder.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:585]: The problem is that during split the first constraint (flat head height <= final_head height) is momentarily broken for children shards :(
Because parent state at its flat head and the parent deltas both end up into the children shards.

File: chain/chain/src/resharding/resharding_actor.rs (hunk 11 above)
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs:24]: Fair enough, I'll change the enum into separate messages

File: chain/chain/src/flat_storage_resharder.rs (hunk 14 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:583]: Hmm I'm not sure we can start flat storage split AFTER split height <= chain head. I mean, we can definitely do that but then the resulting flat head will be beyond chain head because of deltas.
The only way I see to keep flat head <= chain head is to split deltas into deltas instead of split deltas into flat state directly. 
I know it sounds confusing, we can take it offline

Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:583]: On second thought, you were right, this path can no longer happen

File: chain/chain/src/flat_storage_resharder.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:585]: Update: yes `height > final_head.height` should be impossible with current implementation since the task to catchup is starting only if flat head height <= final head height

File: chain/chain/src/resharding/resharding_actor.rs (hunk 11 above)
Comment (Trisfald) [reply chain/chain/src/resharding/resharding_actor.rs]: Done!

File: chain/chain/src/flat_storage_resharder.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:594]: It should be possible. 

One way is to do the same thing I'm doing now for shard catchup. Basically delay the start of the split task until the resharding hash is final.

Another way would be to call `start_resharding` from resharding manager later, when the resharding block is final. It means storing the last resharding hash in DB and going back every block to see check if that block became final.

I don't have a strong opinion but I feel like the first option is easier because it doesn't require read&write of an extra column/field.

[the remark this answers is not in the document]
Comment (shreyan-gupta) [reply chain/chain/src/flat_storage_resharder.rs:594]: Yes, that's what I had suggested [here](https://github.com/near/nearcore/pull/12312#discussion_r1820701423)

I don't think it would be possible to easily integrate approach 2 as we usually call resharding functions on epoch change/shard layout change.

**Comment by Trisfald** [comment]
I have address the most immediate feedback in the PR. 

As discussed offline, major changes such as delaying the split shard task instead of the catchup task will be done in another PR.

File: chain/chain/src/flat_storage_resharder.rs [hunk 29]
```diff
@@ -460,6 +492,182 @@ impl FlatStorageResharder {
...
+                }
+                flat_head_block_hash = chain_store.get_next_block_hash(&flat_head_block_hash)?;
+                if let Some(changes) = store
+                    .get_delta(shard_uid, flat_head_block_hash)
+                    .map_err(|err| Into::<StorageError>::into(err))?
+                {
+                    merged_changes.merge(changes);
+                    store_update.remove_delta(shard_uid, flat_head_block_hash);
+                }
+                // TODO (resharding): if flat_head_block_hash == state sync hash -> do snapshot
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:605]: mini nit for the todo - remove the space otherwise it won't show up when grepping for `TODO(resharding)`

File: chain/chain/src/resharding/event_type.rs [hunk 30]
```diff
@@ -44,15 +42,13 @@ impl ReshardingEventType {
     ///
     /// # Args:
     /// * `next_shard_layout`: the new shard layout
-    /// * `block_hash`: hash of the last block with the shard layout before `next_shard_layout`
-    /// * `prev_block_hash`: hash of the block preceding `block_hash`
+    /// * `resharding_hash`: hash of the last block with the shard layout before `next_shard_layout`
```
Comment (wacban) [inline chain/chain/src/resharding/event_type.rs:45]: nice, it's good to have a name for that

File: chain/chain/src/resharding/resharding_actor.rs [hunk 31]
```diff
@@ -1,23 +1,88 @@
...
 impl Handler<FlatStorageSplitShardRequest> for ReshardingActor {
     fn handle(&mut self, msg: FlatStorageSplitShardRequest) {
-        self.handle_flat_storage_split_shard_request(msg);
+        match msg.resharder.split_shard_task() {
+            FlatStorageReshardingTaskStatus::Successful { .. } => {
+                // All good.
+            }
+            FlatStorageReshardingTaskStatus::Failed => {
+                panic!("impossible to recover from a flat storage split shard failure!")
+            }
```
Comment (wacban) [inline chain/chain/src/resharding/resharding_actor.rs:28]: nit: It may be beneficial to provide some extra info in the Failed and in the panic message. I guess you log it directly before returning failed but still wouldn't hurt.

**Review by wacban — APPROVED** [review]
LGTM
